### PR TITLE
Add maxSpillRunRows config

### PR DESCRIPTION
**Pull request by duanmeng**
We have seen memory allocation failure from the memory allocator
under Prestissimo on Spark workload. The allocation size is very large ~1GB
and the allocation is made from a spilling memory pool. It is likely caused
by extremely large number of rows so the spill run vector needs to allocate
a large amount of memory. To limit the spilling memory usage, we would
better put a limit on the max spill run (a disk spilling config), that is, the
maximum number of rows in a sorted run on disk.

Details in https://github.com/facebookincubator/velox/issues/7760

### CODE REVIEW

File: velox/exec/Spiller.cpp [hunk 1]
```diff
@@ -569,24 +588,28 @@ void Spiller::finalizeSpill() {
...
-    }
+
     // Number of rows to hash and divide into spill partitions at a time.
-    constexpr int32_t kHashBatchSize = 4096;
+    constexpr int32_t kHashBatchSize = 102;
     std::vector<uint64_t> hashes(kHashBatchSize);
     std::vector<char*> rows(kHashBatchSize);
     const bool isSinglePartition = bits_.numPartitions() == 1;
+
+    int32_t totalRows = 0;
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.cpp:605]:
```
    uint64_t totalRows{0};
    bool hasMoreRows{true};
    for (;;) {
      const auto numRows = container_->listRows(
          iterator, rows.size(), );
      if (numRows == 0) {
          hasMoreRows = false;
      }
      ...
    }
    return hasMoreRows;
```

File: velox/exec/Spiller.h [hunk 2]
```diff
@@ -325,6 +330,7 @@ class Spiller {
   memory::MemoryPool* const pool_;
   const HashBitRange bits_;
   const RowTypePtr rowType_;
+  uint64_t maxSpillRunRows_;
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.h:333]: This can be a const?

File: velox/exec/Spiller.h [hunk 3]
```diff
@@ -290,14 +295,14 @@ class Spiller {
   // Prepares spill runs for the spillable data from all the hash partitions.
   // If 'startRowIter' is not null, we prepare runs starting from the offset
   // pointed by 'startRowIter'.
-  void fillSpillRuns(const RowContainerIterator* startRowIter = nullptr);
+  bool fillSpillRuns(RowContainerIterator* startRowIter = nullptr);
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.h:298]: Can you comment the return bool? thanks!

File: velox/exec/Spiller.cpp [hunk 4]
```diff
@@ -486,8 +497,16 @@ void Spiller::spill(const RowContainerIterator* startRowIter) {
 
   markAllPartitionsSpilled();
 
-  fillSpillRuns(startRowIter);
-  runSpill();
+  // TODO：Add for loop as the fillSpillRuns has max rows limit each time.
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.cpp:500]: What's TODO here?

File: velox/exec/Spiller.cpp [hunk 5]
```diff
@@ -486,8 +497,16 @@ void Spiller::spill(const RowContainerIterator* startRowIter) {
...
+  bool lastFill = false;
+  RowContainerIterator iterator;
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.cpp:502]:
```
nit: s/iterator/rowIter/
```

File: velox/exec/Spiller.cpp [hunk 6]
```diff
@@ -486,8 +497,16 @@ void Spiller::spill(const RowContainerIterator* startRowIter) {
...
+  if (startRowIter != nullptr) {
+    iterator = *startRowIter;
+  }
+  while (!lastFill) {
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.cpp:506]:
```
bool noMoreRows{false};
do {
   noMoreRows = fillSpillRuns(&rowIter);
   runSpill();
} while (!noMoreRows);
```

**Review by xiaoxmeng — COMMENTED** [review]
@duanmeng looks good % minors. Thanks!

File: velox/exec/Spiller.cpp [hunk 7]
```diff
@@ -401,7 +411,7 @@ std::unique_ptr<Spiller::SpillStatus> Spiller::writeSpill(int32_t partition) {
   }
 }
 
-void Spiller::runSpill() {
+void Spiller::runSpill(bool lastFill) {
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.cpp:414]:
```
s/lastFill/lastRun/
```

File: velox/exec/Spiller.cpp [hunk 8]
```diff
@@ -486,8 +497,17 @@ void Spiller::spill(const RowContainerIterator* startRowIter) {
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.cpp:497]: Can you also check type_ can't be order by output type?

File: velox/exec/Spiller.cpp [hunk 9]
```diff
@@ -567,24 +587,28 @@ void Spiller::finalizeSpill() {
   finalized_ = true;
 }
 
-void Spiller::fillSpillRuns(const RowContainerIterator* startRowIter) {
+bool Spiller::fillSpillRuns(RowContainerIterator* iterator) {
   checkEmptySpillRuns();
 
+  uint64_t numRows{0};
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.cpp:593]:
```
bool lastRun{false};
```

File: velox/exec/Spiller.cpp [hunk 10]
```diff
@@ -567,24 +587,28 @@ void Spiller::finalizeSpill() {
...
     constexpr int32_t kHashBatchSize = 4096;
     std::vector<uint64_t> hashes(kHashBatchSize);
     std::vector<char*> rows(kHashBatchSize);
     const bool isSinglePartition = bits_.numPartitions() == 1;
+
+    uint64_t totalRows{0};
     for (;;) {
-      auto numRows = container_->listRows(
-          &iterator, rows.size(), RowContainer::kUnlimited, rows.data());
+      numRows = container_->listRows(
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.cpp:606]:
```
const auto numRows = xxx
if (numRows == 0) {
  lastRun = true;
  break;
}
```

File: velox/exec/Spiller.cpp [hunk 11]
```diff
@@ -605,12 +629,16 @@ void Spiller::fillSpillRuns(const RowContainerIterator* startRowIter) {
...
+
+      totalRows += numRows;
+      if (maxSpillRunRows_ > 0 && totalRows >= maxSpillRunRows_) {
         break;
       }
     }
   }
   updateSpillFillTime(execTimeUs);
+
+  return numRows == 0;
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.cpp:641]:
```
return lastRun;
```

File: velox/exec/Spiller.cpp [hunk 12]
```diff
@@ -449,8 +459,9 @@ void Spiller::runSpill() {
     // aggregation output / orderby output spiller, we expect only one spill
     // call to spill all the rows starting from the specified row offset.
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.cpp:460]:
```
add check above the finish file condition?
VELOX_CHECK_EQ(type_ == Spiller::Type::kOrderByOutput, lastRun);
```

File: velox/exec/tests/SpillerTest.cpp [hunk 13]
```diff
@@ -1040,6 +1045,51 @@ class AllTypes : public SpillerTest,
...
+TEST_P(AllTypes, nonSortedSpillFunctions) {
+  struct {
+    uint64_t maxSpillRunRows;
+
+    std::string debugString() const {
+      return fmt::format("maxSpillRunRows {}", maxSpillRunRows);
+    }
+  } testSettings[] = {{101}, {0}};
+
+  for (const auto& testData : testSettings) {
```
Comment (xiaoxmeng) [inline velox/exec/tests/SpillerTest.cpp:1057]: Do we need this test? I assume non-sorted spiller will ignore the max spill run parameter except hash join spiller type? Can we just verify that the number of files spilled by hash join spiller is controlled by this parameter? Like targetFileSize is extreme large?

File: velox/exec/tests/SpillerTest.cpp [hunk 14]
```diff
@@ -1395,6 +1424,67 @@ TEST_P(OrderByOutputOnly, basic) {
...
+      ASSERT_EQ(numFiles, 1);
+      ASSERT_EQ(spillPartitionSet.size(), 1);
+    } else if (
+        type_ == Spiller::Type::kAggregateInput ||
+        type_ == Spiller::Type::kOrderByInput) {
+      // Need sort.
+      ASSERT_EQ(numFiles, testData.expectedNumFiles);
+      ASSERT_EQ(spillPartitionSet.size(), 1);
+    } else {
+      ASSERT_GT(numFiles, 1);
```
Comment (xiaoxmeng) [inline velox/exec/tests/SpillerTest.cpp:1483]: If we set target file to an extreme large value, then the number of files will be 1 per each partition or at least smaller than the number of partitions? To verify that, we can have a test setup the number of partitions is one and four?

File: velox/exec/Spiller.cpp (hunk 12 above)
Comment (duanmeng) [reply velox/exec/Spiller.cpp:460]: Added

File: velox/exec/tests/SpillerTest.cpp (hunk 14 above)
Comment (duanmeng) [reply velox/exec/tests/SpillerTest.cpp]: Got it, I've made the target file size uint64_t max to force only one file per partition for the HashBuild spiller.
```C++
setupSpiller(std::numeric_limits<uint64_t>::max(),
   ...
} else {
      ASSERT_EQ(numFiles, hashBits_.numPartitions());
      ASSERT_EQ(spillPartitionSet.size(), numFiles);
    }
```

File: velox/core/QueryConfig.h [hunk 15]
```diff
@@ -234,6 +234,13 @@ class QueryConfig {
   static constexpr const char* kOrderBySpillMemoryThreshold =
       "order_by_spill_memory_threshold";
 
+  /// The max row numbers to fill and spill for each spill run. This is used to
+  /// cap the memory used for spilling. If it is zero, then there is no limit
+  /// and spilling might run out of memory.
+  /// According to the test, the recommendation value is 12 million rows, which
+  /// may roughly use 128 MB of memory when a fill spill runs.
```
Comment (xiaoxmeng) [inline velox/core/QueryConfig.h:241]:
```
s/when a fill spill runs./when fills a spill run/
```

File: velox/exec/Spiller.h [hunk 16]
```diff
@@ -61,6 +61,7 @@ class Spiller {
       common::CompressionKind compressionKind,
       memory::MemoryPool* pool,
       folly::Executor* executor,
+      uint64_t maxSpillRunRows = 0,
```
Comment (xiaoxmeng) [inline velox/exec/Spiller.h:64]: Can we require caller to pass maxSpillRunRows but for some constructor which always takes zero, then let's remove it from ctor? Thanks!

File: velox/exec/tests/SpillerTest.cpp (hunk 13 above)
Comment (xiaoxmeng) [reply velox/exec/tests/SpillerTest.cpp:1057]: Can you add to test a large value per our discussion? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@duanmeng some nits. Thanks!

File: velox/core/QueryConfig.h [hunk 17]
```diff
@@ -234,6 +234,13 @@ class QueryConfig {
   static constexpr const char* kOrderBySpillMemoryThreshold =
       "order_by_spill_memory_threshold";
 
+  /// The max row numbers to fill and spill for each spill run. This is used to
```
Comment (xiaoxmeng) [inline velox/core/QueryConfig.h:237]: Can you update configs.rst? Thanks!

File: velox/exec/tests/SpillerTest.cpp [hunk 18]
```diff
@@ -1246,20 +1262,34 @@ TEST_P(AggregationOutputOnly, basic) {
...
+      {5'000 * 2, 100'000}};
 
   for (const auto& testData : testSettings) {
     SCOPED_TRACE(testData.debugString());
 
     setupSpillData(rowType_, numKeys_, numRows, 0);
     sortSpillData();
     // NOTE: target file size is ignored by aggregation output spiller type.
-    setupSpiller(0, 1'000'000, 0, false);
+    setupSpiller(0, 1'000'000, 0, false, testData.maxSpillRunRows);
```
[the remark this answers is not in the document]
Comment (duanmeng) [reply velox/exec/tests/SpillerTest.cpp:1301]: We can not check the row number per our discussion.

File: velox/core/QueryConfig.h [hunk 19]
```diff
@@ -237,6 +237,13 @@ class QueryConfig {
...
+  /// cap the memory used for spilling. If it is zero, then there is no limit
+  /// and spilling might run out of memory.
+  /// According to the test, the recommendation value is 12 million rows, which
```
Comment (xiaoxmeng) [inline velox/core/QueryConfig.h:243]:
```
/// Based on offline test results, the default value is set to 12 million rows which uses ~128MB memory ...
```

File: velox/docs/configs.rst [hunk 20]
```diff
@@ -278,6 +278,16 @@ Spilling
        spilling which might use recursive spilling when the build table is very large. -1 means unlimited.
        In this case an extremely large query might run out of spilling partition bits. The max spill level
        can be used to prevent a query from using too much io and cpu resources.
+   * - max_spill_run_rows
+     - integer
+     - 12582912
+     - The max row numbers to fill and spill for each spill run. This is used to cap the memory used for spilling. If it
```
Comment (xiaoxmeng) [inline velox/docs/configs.rst:284]:
```
The max number of rows to fill and spill for ...
```

File: velox/docs/configs.rst [hunk 21]
```diff
@@ -278,6 +278,16 @@ Spilling
...
+       is zero, then there is no limit and spilling might run out of memory. According to the test, the recommendation
```
Comment (xiaoxmeng) [inline velox/docs/configs.rst:285]:
```
Based on the offline test result, the recommended value is 12 ... use ... to fill a spill run. The relation between the number of spill rows and the corresponding memory usage are as follows:
```

File: velox/core/QueryConfig.h (hunk 19 above)
Comment (duanmeng) [reply velox/core/QueryConfig.h]: Fixed

File: velox/core/QueryConfig.h [hunk 22]
```diff
@@ -237,6 +237,13 @@ class QueryConfig {
   static constexpr const char* kOrderBySpillMemoryThreshold =
       "order_by_spill_memory_threshold";
 
+  /// The max row numbers to fill and spill for each spill run. This is used to
+  /// cap the memory used for spilling. If it is zero, then there is no limit
+  /// and spilling might run out of memory.
+  /// Based on offline test results, the default value is set to 12 million rows
+  /// which uses ~128MB memory when fills a spill run.
```
Comment (xiaoxmeng) [inline velox/core/QueryConfig.h:244]:
```
s/when fills/to fill/
```

File: velox/docs/configs.rst [hunk 23]
```diff
@@ -278,6 +278,16 @@ Spilling
        spilling which might use recursive spilling when the build table is very large. -1 means unlimited.
        In this case an extremely large query might run out of spilling partition bits. The max spill level
        can be used to prevent a query from using too much io and cpu resources.
+   * - max_spill_run_rows
+     - integer
+     - 12582912
+     - The max number of rows to fill and spill for each spill run. This is used to cap the memory used for spilling.
+       If it is zero, then there is no limit and spilling might run out of memory. Based on offline test results, the
+       default value is set to 12 million rows which uses ~128MB memory when fills a spill run.
```
Comment (xiaoxmeng) [inline velox/docs/configs.rst:286]:
```
s/when fills/to fill/
```

**Pull request closed without merging**